core/src/main/java/org/apache/accumulo/core/clientImpl/TabletServerBatchReader.java

```diff
@@ -72,16 +72,15 @@ protected TabletServerBatchReader(ClientContext context, Class<?> scopeClass, Ta
 
     queryThreadPool = ThreadPools.getFixedThreadPool(numQueryThreads,
         "batch scanner " + batchReaderInstance + "-", false);
+    // Call shutdown on this thread pool in case the caller does not call close().
     cleanable = CleanerUtil.shutdownThreadPoolExecutor(queryThreadPool, log);
   }
 
   @Override
   public void close() {
     if (closed.compareAndSet(false, true)) {
-      // deregister cleanable, but it won't run because it checks
-      // the value of closed first, which is now true
-      cleanable.clean();
       queryThreadPool.shutdownNow();
+      cleanable.clean();
```

Comment:
We really should deregister the cleanable first, so we don't get duplicate exceptions from that if there's a problem calling the shutdown.

Reply:
So we really only need `cleanable.clean()` here as it deregisters *and* invokes the cleaning action.

Reply:
In the next commit I left the order of calls the same. If we only use `cleanable.clean()` or put `cleanable.clean()` first, then we will always get the WARN log message: ThreadPoolExecutor found unreferenced without calling shutdown() or shutdownNow().

Reply:
> In the next commit I left the order of calls the same. If we only use `cleanable.clean()` or put `cleanable.clean()` first, then we will always get the WARN log message: ThreadPoolExecutor found unreferenced without calling shutdown() or shutdownNow().

If shutdown is only called in the close method, then it can be guarded with the closed variable that you are now passing to the cleaner. I guess what I don't know is if we ever want to call shutdown while leaving the containing object unclosed.

Reply:
The reason to ensure the cleanable is deregistered first, is because the `shutdownNow()` could throw an exception, causing the cleanable to stick around until the object is garbage collected. In general, we don't want cleanables to stick around if they are going to be NOOPs later (the AtomicBoolean "closed" variable will ensure it's a NOOP, so we don't try to close it both inside the cleanable and outside).